calico_containers/adapter/datastore.py

```diff
@@ -794,12 +914,28 @@ def get_endpoints(self, hostname, container_id):
         # Extract all of the endpoints.
         eps = []
         for endpoint in endpoints:
-            (_, _, _, _, _, _, _, _, _, endpoint_id) = \
-                                                 endpoint.key.split("/", 9)
-            eps.append(Endpoint.from_json(endpoint_id, endpoint.value))
+            eps.append(Endpoint.from_json(endpoint.key, endpoint.value))
         return eps
 
     @handle_errors
+    def get_endpoint_from_id(self, endpoint_id):
+        """
+        Get the endpoint specified by the endpoint id.
+
+        :param endpoint_id: The endpoint ID.
+        :return:  a tuple of: (raw json, Endpoint)
+        """
+        leaves = self.etcd_client.read(HOSTS_PATH, recursive=True).leaves
+
+        # Extract all of the endpoints.
+        for leaf in leaves:
+            ep = Endpoint.from_json(leaf.key, leaf.value)
+            if ep and ep.endpoint_id == endpoint_id:
+                return ep
+        raise KeyError("Endpoint with ID %s was not found." %
+                       endpoint_id)
+
+    @handle_errors
     def get_hosts(self):
```

Comment:
WE should retire this.  It's badly named (it's actually getting the endpoints) and the function is effectively duplicated by query_endpoints.